contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,479 @@
+
+  state->handleHeaderGolangStatus(static_cast<TcpUpstreamStatus>(go_status));
+
+  bool send_data_to_upstream = (state->filterState() == FilterState::Done);
+
+  trySendProxyData(send_data_to_upstream, end_stream);
+
+  if (send_data_to_upstream) {
+    // when go side decide to send data to upstream in encodeHeaders, here will directly send and encodeData will not send.
+    state->doDataList.moveOut(buf);
+    upstream_conn_data_->connection().write(buf, upstream_conn_self_half_close_);
+  }
+
+  return Envoy::Http::okStatus();
+}
+
+void TcpUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream encodeData, data length: {}, end_stream: {}", data.length(), end_stream);
+
+  ProcessorState* state = encoding_state_;
+
+  switch (state->filterState()) {
+  case FilterState::WaitingData:
+    encodeDataGo(state, data, end_stream);
+    break;
+
+  case FilterState::WaitingAllData:
+    if (end_stream) {
+      if (!state->isBufferDataEmpty()) {
+        // NP: new data = data_buffer_ + data
+        state->addBufferData(data);
+        data.move(state->getBufferData());
+      }
+      encodeDataGo(state, data, end_stream);
+    } else {
+      ENVOY_LOG(debug, "tcp upstream encodeData, appending data to buffer");
+      state->addBufferData(data);
+    }
+    break;
+
+  case FilterState::Done:
+    ENVOY_LOG(debug, "tcp upstream encodeData, already send data to upstream");
+    return;
+
+  default:
+    ENVOY_LOG(error, "tcp upstream encodeData, unexpected state: {}", state->stateStr());
+    PANIC("unreachable");
+    break;
+  }
+}
+
+// TODO(duxin40): use golang to convert trailers to data buffer.
+void TcpUpstream::encodeTrailers(const Envoy::Http::RequestTrailerMap&) {
+  Buffer::OwnedImpl data;
+  upstream_conn_data_->connection().write(data, true);
+}
+
+void TcpUpstream::readDisable(bool disable) {
+  if (upstream_conn_data_->connection().state() != Network::Connection::State::Open) {
+    return;
+  }
+  upstream_conn_data_->connection().readDisable(disable);
+}
+
+void TcpUpstream::resetStream() {
+  upstream_request_ = nullptr;
+  upstream_conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+}
+
+void TcpUpstream::onUpstreamData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream onUpstreamData, data length: {}, end_stream: {}", data.length(), end_stream);
+
+  initResponse();
+
+  DecodingProcessorState* state = decoding_state_;
+  state->resp_headers = static_cast<Envoy::Http::RequestOrResponseHeaderMap*>(resp_headers_.get());
+  Buffer::Instance& buffer = state->doDataList.push(data);
+  auto s = dynamic_cast<processState*>(state);
+  state->setFilterState(FilterState::ProcessingData);
+
+  GoUint64 go_status = dynamic_lib_->envoyGoOnUpstreamData(
+    s, end_stream ? 1 : 0, (*resp_headers_).size(), (*resp_headers_).byteSize(), reinterpret_cast<uint64_t>(&buffer), buffer.length());
+  state->doDataList.moveOut(data);
```

Comment:
doDataList is useless? seems we always clear it here.

Reply:
done